### PR TITLE
Feature/fixing clang gnu compiler warnings

### PR DESCRIPTION
**Pull request by phelter**
Fixing majority of the clang compiler warnings.

Description
-----------

Fixed majority of clang and gnu compiler warnings.  Primary ones are:
* misuse of doxygen single line comments
* replaced wrong type usage for Interrupt return and clear interrupt macros.

List of compiler warnings goes from:

```cmake
target_compile_options( freertos_kernel
  PRIVATE
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-cast-align>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-covered-switch-default>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-documentation>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-extra-semi-stmt>
    $<$<COMPILE_LANG_AND_ID:C,GNU>:-Wno-int-to-pointer-cast>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-missing-noreturn>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-missing-variable-declarations>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-padded>
    $<$<COMPILE_LANG_AND_ID:C,GNU>:-Wno-pointer-to-int-cast>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-shorten-64-to-32>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-sign-conversion>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-unused-macros>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wno-unused-variable>
)
# For linux port
target_compile_options( freertos_kernel_port
  PRIVATE
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-disabled-macro-expansion>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-documentation>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-missing-noreturn>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-missing-variable-declarations>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-padded>
    $<$<COMPILE_LANG_AND_ID:C,GNU>:-Wno-pointer-to-int-cast>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-reserved-macro-identifier>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-sign-conversion>
    $<$<COMPILE_LANG_AND_ID:C,GNU>:-Wno-type-limits>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-unused-macros>
)
```

To:
```cmake
target_compile_options( freertos_kernel
  PRIVATE
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-cast-align>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-covered-switch-default>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-missing-variable-declarations>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-padded>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-unused-macros>
)
target_compile_options( freertos_kernel_port
  PRIVATE
     $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-disabled-macro-expansion>
     $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-padded>
     $<$<COMPILE_LANG_AND_ID:C,Clang>:-Wno-reserved-macro-identifier>
)
```

Test Steps
-----------
Compile with:

```cmake
add_compile_options(
    ### C Based Language and ID.
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wall>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wextra>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wpedantic>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Werror>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Weverything>
)
```

Related Issue
-----------
Tried to have this list in the CMakeLists.txt originally in PR #571, but was overruled, so reducing list for future checks.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/timers.h [hunk 1]
```diff
@@ -1361,6 +1361,20 @@ BaseType_t xTimerGenericCommand( TimerHandle_t xTimer,
 
 #endif
 
+#if  ( configUSE_DAEMON_TASK_STARTUP_HOOK != 0 )
```
Comment (aggarg) [inline include/timers.h:1364]:
```suggestion
#if ( configUSE_DAEMON_TASK_STARTUP_HOOK != 0 )
```

File: include/timers.h [hunk 2]
```diff
@@ -1361,6 +1361,20 @@ BaseType_t xTimerGenericCommand( TimerHandle_t xTimer,
...
+/**
+ *  timers.h
+ * @code{c}
+ * void vApplicationTickHook( void );
+ * @endcode
+ *
+ * This hook function is called in the timers task before task executes.
+ */
+    void vApplicationDaemonTaskStartupHook( void );
+
```
Comment (aggarg) [inline include/timers.h:1375]:
```suggestion
/**
 * timers.h
 * @code{c}
 * void vApplicationTickHook( void );
 * @endcode
 *
 * This hook function is called form the timer task once when the task starts running.
 */
    void vApplicationDaemonTaskStartupHook( void );

```

File: portable/MemMang/heap_4.c [hunk 3]
```diff
@@ -411,7 +411,15 @@ static void prvHeapInit( void ) /* PRIVILEGED_FUNCTION */
 
     /* To start with there is a single free block that is sized to take up the
      * entire heap space, minus the space taken by pxEnd. */
+    #if defined( __clang__ )
```
Comment (aggarg) [inline portable/MemMang/heap_4.c:414]: Can we keep the compiler specific changes out of the common files? Is it possible to suppress this warning at the file level in the application CMakeFile?

File: portable/ThirdParty/GCC/Posix/port.c [hunk 4]
```diff
@@ -141,7 +144,7 @@ portSTACK_TYPE * pxPortInitialiseStack( portSTACK_TYPE * pxTopOfStack,
      */
     thread = ( Thread_t * ) ( pxTopOfStack + 1 ) - 1;
     pxTopOfStack = ( portSTACK_TYPE * ) thread - 1;
-    ulStackSize = ( pxTopOfStack + 1 - pxEndOfStack ) * sizeof( *pxTopOfStack );
+    ulStackSize = (size_t)( pxTopOfStack + 1 - pxEndOfStack ) * sizeof( *pxTopOfStack );
```
Comment (aggarg) [inline portable/ThirdParty/GCC/Posix/port.c:147]:
```suggestion
    ulStackSize = ( size_t )( pxTopOfStack + 1 - pxEndOfStack ) * sizeof( *pxTopOfStack );
```

File: portable/ThirdParty/GCC/Posix/port.c [hunk 5]
```diff
@@ -340,7 +343,7 @@ static uint64_t prvGetTimeNs( void )
 
     clock_gettime( CLOCK_MONOTONIC, &t );
 
-    return t.tv_sec * 1000000000ULL + t.tv_nsec;
+    return (uint64_t)t.tv_sec * 1000000000ULL + (uint64_t)t.tv_nsec;
```
Comment (aggarg) [inline portable/ThirdParty/GCC/Posix/port.c:346]:
```suggestion
    return ( uint64_t ) t.tv_sec * 1000000000ULL + ( uint64_t ) t.tv_nsec;
```

File: portable/ThirdParty/GCC/Posix/port.c [hunk 6]
```diff
@@ -115,6 +115,9 @@ static void prvPortYieldFromISR( void );
 /*-----------------------------------------------------------*/
 
 static void prvFatalError( const char * pcCall,
+                           int iErrno ) __attribute__ ((__noreturn__));
+
```
Comment (aggarg) [inline portable/ThirdParty/GCC/Posix/port.c:119]:
```suggestion
static void prvFatalError( const char * pcCall,
                           int iErrno ) __attribute__ ((__noreturn__));
/*-----------------------------------------------------------*/

```

File: stream_buffer.c [hunk 7]
```diff
@@ -371,6 +371,11 @@ static void prvInitialiseNewStreamBuffer( StreamBuffer_t * const pxStreamBuffer,
             pucAllocatedMemory = NULL;
         }
 
+        #if defined( __clang__ )
```
Comment (aggarg) [inline stream_buffer.c:374]: Same suggestion as above about keeping the compiler specific changes out of the common files.

File: stream_buffer.c [hunk 8]
```diff
@@ -1372,9 +1379,9 @@ static void prvInitialiseNewStreamBuffer( StreamBuffer_t * const pxStreamBuffer,
         /* The value written just has to be identifiable when looking at the
          * memory.  Don't use 0xA5 as that is the stack fill value and could
          * result in confusion as to what is actually being observed. */
-        const BaseType_t xWriteValue = 0x55;
-        configASSERT( memset( pucBuffer, ( int ) xWriteValue, xBufferSizeBytes ) == pucBuffer );
-    } /*lint !e529 !e438 xWriteValue is only used if configASSERT() is defined. */
+    #define STREAM_BUFFER_BUFFER_WRITE_VALUE    ( 0x55 )
```
Comment (aggarg) [inline stream_buffer.c:1382]: What is the reason for this change?

File: tasks.c [hunk 9]
```diff
@@ -3426,7 +3429,9 @@ void vTaskMissedYield( void )
  * void prvIdleTask( void *pvParameters );
  *
  */
-static portTASK_FUNCTION( prvIdleTask, pvParameters )
+static void prvIdleTask( void * pvParameters ) __attribute__( ( __noreturn__ ) );
```
Comment (aggarg) [inline tasks.c:3432]: This change would probably not work for all compilers? If so, we need to remove it.

File: timers.c [hunk 10]
```diff
@@ -565,7 +565,7 @@
     }
 /*-----------------------------------------------------------*/
 
-    static portTASK_FUNCTION( prvTimerTask, pvParameters )
+    static portTASK_FUNCTION( prvTimerTask, pvParameters ) __attribute__( ( __noreturn__ ) );
```
Comment (aggarg) [inline timers.c:568]: Same suggestion as in tasks.c about portability.

**Comment by RichardBarry** [comment]
Looks like this is in hang already, some notes all the same:
1) We can't have any compiler specific code in common files, only in the port layer.  Compiler specific code includes preprocessor macros that test which compiler is in use.
2) Make sure updates comment block characters till work with Doxygen.

File: include/timers.h (hunk 1 above)
Comment (phelter) [reply include/timers.h:1364]: Suggest using `clang-format` instead of `uncrustify` if you want to enforce that level of formatting requirements.

Comment (phelter) [reply include/timers.h:1364]: Changed.

File: include/timers.h (hunk 2 above)
Comment (phelter) [reply include/timers.h:1375]: Changed.

File: portable/MemMang/heap_4.c (hunk 3 above)
Comment (phelter) [reply portable/MemMang/heap_4.c:414]: Yes it is possible however it means that the check is disabled for the entire target and/or the entire heap_4.c file instead of just this one file location.  Moved to CMakeFile as part of freertos_kernel target.

File: portable/ThirdParty/GCC/Posix/port.c (hunk 4 above)
Comment (phelter) [reply portable/ThirdParty/GCC/Posix/port.c:147]: Changed, but if you require this level of formatting compliance, please incorporate it into `uncrustify` definition and/or use `clang-format` which has this level of formatting definition

File: portable/ThirdParty/GCC/Posix/port.c (hunk 5 above)
Comment (phelter) [reply portable/ThirdParty/GCC/Posix/port.c:346]: See Comment: https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/620#discussion_r1104071411

File: stream_buffer.c (hunk 7 above)
Comment (phelter) [reply stream_buffer.c:374]: Moved to CMakeFile.txt

File: portable/ThirdParty/GCC/Posix/port.c (hunk 6 above)
Comment (phelter) [reply portable/ThirdParty/GCC/Posix/port.c:119]: Added.

File: stream_buffer.c (hunk 8 above)
Comment (phelter) [reply stream_buffer.c:1382]: Part of it was a merge conflict when merging in main.  I've fixed that.  Change. there is no need to have the xWriteValue as a const variable. can use a # define instead and removes the lint !e529 and !e438 issue of not using xWriteValue when configAssert is not defined.

File: tasks.c (hunk 9 above)
Comment (phelter) [reply tasks.c]: Added `portNORETURN` to all of the known GNU compiler ports and the `FreeRTOS.h` file if not defined.  This will solve the issue of not being supported on some compilers.  Note for C23 - noreturn is supported.

For IAR it is `__noreturn` based on: https://wwwfiles.iar.com/arm/webic/doc/EWARM_DevelopmentGuide.ENU.pdf
For CCS did not look into it.

But would suggest you specify compiler specific defines in one location at the top of each `portable/GCC` `portable/IAR`, etc. so that they are defined once and included in each of the portable.h files under the respective compiler.  That way you are guaranteed not to miss one in one or some of the ports (which is the case for some of these values at the moment).

But out of scope of this PR.

**Comment by phelter** [comment]
> Looks like this is in hang already, some notes all the same:
> 
>     1. We can't have any compiler specific code in common files, only in the port layer.  Compiler specific code includes preprocessor macros that test which compiler is in use.

Fixed with latest commit.

>     2. Make sure updates comment block characters till work with Doxygen.

This can be done by running a clang compile with the following compile options:

```cmake
add_compile_options(
    ### C Based Language and ID.
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wall>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wextra>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Wpedantic>
    $<$<COMPILE_LANG_AND_ID:C,Clang,GNU>:-Werror>
    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Weverything>
)
```

the -Wdocumentation-* checks validity of doxygen information - i.e. invalid or missing params, and wrong usage of doxygen.  You can also run Doxygen on the entire suite of C files using CMake and have warnings as errors defined when attempting to create the documentation as one of the build steps.

File: portable/MemMang/heap_4.c (hunk 3 above)
Comment (aggarg) [reply portable/MemMang/heap_4.c:414]: So, can this be removed now?

File: stream_buffer.c (hunk 7 above)
Comment (aggarg) [reply stream_buffer.c:374]: Can this be removed then?

File: timers.c [hunk 11]
```diff
@@ -565,7 +565,7 @@
     }
 /*-----------------------------------------------------------*/
 
-    static portTASK_FUNCTION( prvTimerTask, pvParameters )
+    static portTASK_FUNCTION( prvTimerTask, pvParameters ) portNORETURN;
```
Comment (aggarg) [inline timers.c:568]: This does not seem correct?

File: portable/ThirdParty/GCC/Posix/port.c (hunk 5 above)
Comment (aggarg) [reply portable/ThirdParty/GCC/Posix/port.c:346]: This also I do not see.

**Comment by aggarg** [comment]
Some checks are failing likely because of the syntax error in timers.c. Please look into the failed checks.

Thank you again for your contribution.

File: timers.c (hunk 11 above)
Comment (phelter) [reply timers.c]: What is `this` ? What is not correct about it?

File: portable/MemMang/heap_4.c (hunk 3 above)
Comment (phelter) [reply portable/MemMang/heap_4.c]: Yes removed - updated list in description to include the cast-align in cmake.

**Comment by phelter** [comment]
@RichardBarry and @aggarg - the changes requested have all been performed and the unit tests now pass.  I'd suggest adding a build with clang and would also suggest using clang-tidy on this code base as well.

**Comment by phelter** [comment]
Ping @RichardBarry and @aggarg .  Would like to no longer maintain the fork that I have in use, so please approve at your earliest convenience.

File: tasks.c [hunk 12]
```diff
@@ -3426,7 +3429,8 @@ void vTaskMissedYield( void )
  * void prvIdleTask( void *pvParameters );
  *
  */
-static portTASK_FUNCTION( prvIdleTask, pvParameters )
+
+portTASK_FUNCTION( prvIdleTask, pvParameters )
```
Comment (phelter) [inline tasks.c:3433]: Note code coverage for this line is not expected.  There is no change except to remove the `static` from the `function definition` because it is already defined in the `function declaration`.  Please waive.

File: queue.c [hunk 13]
```diff
@@ -1209,7 +1209,7 @@ BaseType_t xQueueGiveFromISR( QueueHandle_t xQueue,
                               BaseType_t * const pxHigherPriorityTaskWoken )
 {
     BaseType_t xReturn;
-    UBaseType_t uxSavedInterruptStatus;
+    portBASE_TYPE xSavedInterruptStatus;
```
Comment (phelter) [inline queue.c:1212]: Note only changing the type here because the original had an incorrect type.  These are defined by the `port types - portBaseType`` and not by the `kernel types - UBaseType_t`.  Please waive.

**Comment by aggarg** [comment]
> Ping @RichardBarry and @aggarg . Would like to no longer maintain the fork that I have in use, so please approve at your earliest convenience.

Apologies for the delay. Will get it done today.

**Review by aggarg — APPROVED** [review]

File: stream_buffer.c [hunk 14]
```diff
@@ -70,7 +70,7 @@
             ( pxStreamBuffer )->xTaskWaitingToSend = NULL;                \
         }                                                                 \
     }                                                                     \
-    ( void ) xTaskResumeAll();
+    ( void ) xTaskResumeAll()
```
Comment (Skptak) [inline stream_buffer.c:73]: Is this missing the semi-colon here?

**Review by Skptak — APPROVED** [review]

**Comment by aggarg** [comment]
Thank you for your contribution.

File: CMakeLists.txt [hunk 15]
```diff
@@ -222,6 +222,47 @@ elseif((FREERTOS_PORT STREQUAL "A_CUSTOM_PORT") AND (NOT TARGET freertos_kernel_
     "       freertos_kernel)")
 endif()
 
+########################################################################
+# Requirements
+set(CMAKE_C_STANDARD 99)
```
Comment (aggarg) [inline CMakeLists.txt:227]: @phelter  Why do you need C99? We want to support C89. Did you notice anything C99 specific in the kernel?

Comment (aggarg) [reply CMakeLists.txt:227]: Thank you for the clarification. I did build the kernel with `-Wdeclaration-after-statement` and I see no warning from kernel source. I will remove the following 2 lines which were added in this PR - 
```
set(CMAKE_C_STANDARD 99)
set(CMAKE_C_STANDARD_REQUIRED ON)
```

Comment (aggarg) [reply CMakeLists.txt:227]: Here is the PR - https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/633
